File: javascript/say/say.js
const numbersToWords = {
  0: 'zero', 1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five', 6: 'six', 7: 'seven', 8: 'eight', 9: 'nine', 10: 'ten',
  11: 'eleven', 12: 'twelve', 13: 'thirteen', 14: 'fourteen', 15: 'fifteen', 16: 'sixteen', 17: 'seventeen', 18: 'eighteen', 19: 'nineteen',
  20: 'twenty', 30: 'thirty', 40: 'forty', 50: 'fifty', 60: 'sixty', 70: 'seventy', 80: 'eighty', 90: 'ninety',
};

const scales = ['', 'thousand', 'million', 'billion']; // No trillion needed as per instructions (max 999,999,999,999)

// Helper function to convert a number less than 1000 to words
const convertChunk = (num) => {
  if (num === 0) return '';

  let words = '';

  if (num >= 100) {
    words += numbersToWords[Math.floor(num / 100)] + ' hundred';
    num %= 100;
    if (num > 0) {
      words += ' ';
    }
  }

  if (num > 0) {
    if (num < 20) {
      words += numbersToWords[num];
    } else {
      words += numbersToWords[Math.floor(num / 10) * 10];
      if (num % 10 > 0) {
        words += '-' + numbersToWords[num % 10];
      }
    }
  }
  return words;
};

export const say = (n) => {
  if (n < 0 || n > 999999999999) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  if (n === 0) {
    return numbersToWords[0];
  }

  let words = '';
  let chunkIndex = 0;

  while (n > 0) {
    if (n % 1000 !== 0) {
      const chunkWords = convertChunk(n % 1000);
      const scaleWord = scales[chunkIndex];
      words = chunkWords + (scaleWord ? ' ' + scaleWord : '') + (words ? ' ' + words : '');
    }
    n = Math.floor(n / 1000);
    chunkIndex++;
  }

  return words.trim();
};

File: go/say/say.go
package say

func Say(n int64) (string, bool) {
	if n < 0 || n >= 1_000_000_000_000 {
		return "", false
	}
	if n == 0 {
		return "zero", true
	}

	var parts []string

	billions := n / 1_000_000_000
	if billions > 0 {
		parts = append(parts, sayBelowThousand(billions)+" billion")
		n %= 1_000_000_000
	}

	millions := n / 1_000_000
	if millions > 0 {
		parts = append(parts, sayBelowThousand(millions)+" million")
		n %= 1_000_000
	}

	thousands := n / 1_000
	if thousands > 0 {
		parts = append(parts, sayBelowThousand(thousands)+" thousand")
		n %= 1_000
	}

	if n > 0 {
		parts = append(parts, sayBelowThousand(n))
	}

	return joinParts(parts), true
}

func sayBelowThousand(n int64) string {
	var parts []string

	hundreds := n / 100
	if hundreds > 0 {
		parts = append(parts, digitToWord(hundreds)+" hundred")
		n %= 100
	}

	if n > 0 {
		parts = append(parts, sayBelowHundred(n))
	}

	return joinParts(parts)
}

func sayBelowHundred(n int64) string {
	if n < 20 {
		return digitToWord(n)
	}

	tens := n / 10
	ones := n % 10

	tensWord := tensToWord(tens * 10)
	if ones == 0 {
		return tensWord
	}
	return tensWord + "-" + digitToWord(ones)
}

func digitToWord(n int64) string {
	switch n {
	case 0:
		return "zero"
	case 1:
		return "one"
	case 2:
		return "two"
	case 3:
		return "three"
	case 4:
		return "four"
	case 5:
		return "five"
	case 6:
		return "six"
	case 7:
		return "seven"
	case 8:
		return "eight"
	case 9:
		return "nine"
	case 10:
		return "ten"
	case 11:
		return "eleven"
	case 12:
		return "twelve"
	case 13:
		return "thirteen"
	case 14:
		return "fourteen"
	case 15:
		return "fifteen"
	case 16:
		return "sixteen"
	case 17:
		return "seventeen"
	case 18:
		return "eighteen"
	case 19:
		return "nineteen"
	default:
		return ""
	}
}

func tensToWord(n int64) string {
	switch n {
	case 20:
		return "twenty"
	case 30:
		return "thirty"
	case 40:
		return "forty"
	case 50:
		return "fifty"
	case 60:
		return "sixty"
	case 70:
		return "seventy"
	case 80:
		return "eighty"
	case 90:
		return "ninety"
	default:
		return ""
	}
}

func joinParts(parts []string) string {
	result := ""
	for i, part := range parts {
		if i > 0 {
			result += " "
		}
		result += part
	}
	return result
}
